File: Unit 01 - Linear Classifiers and Generalizations/recitation01_code_tumorDiagnosisDemo.jl
# Scikit-Learn:
#  - Python: https://scikit-learn.org/stable/
#  - Julia port/wrapper: https://github.com/cstjean/ScikitLearn.jl

# Import libraries
using Statistics, StatsPlots, ScikitLearn
using ScikitLearn.CrossValidation: cross_val_score
@sk_import datasets: load_breast_cancer
@sk_import preprocessing : scale
@sk_import linear_model: SGDClassifier

# Load and prepare data
cancer_data = load_breast_cancer()
y = cancer_data["target"] # Training labels ('malignant = 0', 'benign = 1')
X = cancer_data["data"] # 30 attributes; https://scikit-learn.org/stable/datasets/index.html#breast-cancer-dataset
X = scale(X) # scale each data attribute to zero-mean and unit variance

# Plot the first 2 attributes of training points
colors = [y == 1 ? "blue" : "red" for y in y]
labels = [y == 1 ? "benign" : "malign" for y in y]
scatter(X[:,1],X[:,2], colour=colors, title="Classified tumors",xlabel="Tumor Radius", ylabel="Tumor Texture", group=labels)


# Compute cross-validation scores
alpha = eps():0.005:1 # Range of hyperparameter values 1E-15 to 1 by 0.005
val_scores = zeros(length(alpha)) # Initialize validation score for each alpha value
for (i,a) in enumerate(alpha) # for each alpha value
    # Set up SVM with hinge loss and l2 norm regularization
    model = SGDClassifier(loss="hinge", penalty="l2", alpha=a)
    # Calculate cross validation scores for 5-fold cross-validation
    score = cross_val_score(model, X, y; cv=5)
    val_scores[i] = mean(score) # Calculate mean of the 5 scores
end

# Plot how cross-validation score changes with alpha
plot(alpha,val_scores,xlims=(0,1), xlabel="alpha",ylabel="Mean Cross-Validation Accuracy",label="")

# Determine the alpha that maximizes the cross-validation score
ind = argmax(val_scores)
alpha_star = alpha[ind]
println("alpha_star = $alpha_star")
vline!([alpha_star], label="alpha*")

# Train model with alpha_star
model_star = SGDClassifier(loss="hinge", penalty="l2", alpha=alpha_star)
model_trained = model_star.fit(X,y)
print("Training Accuracy = $(model_trained.score(X,y))")

# Plot decision boundary of trained model
slope = model_trained.coef_[1,2]/-model_trained.coef_[1,1]
x1 = -10:0.5:10
y1 = slope*x1
scatter(X[:,1],X[:,2], colour=colors, title="Classified tumors",xlims=(-4,4),ylims=(-6,6), xlabel="Tumor Radius", ylabel="Tumor Texture", group=labels)
plot!(x1,y1, linestyle = :dot, label="classifier")

File: Unit 01 - Linear Classifiers and Generalizations/recitation01_code_tumorDiagnosisDemo.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Tumor diagnosis demo
Created on Fri Feb 28 16:08:45 2020

@author: MITx micromaster
"""




### Scikit-Learn: https://scikit-learn.org/stable/


# In[21]:


# Imports
import numpy as np
import matplotlib.pyplot as plt # for plotting 
import seaborn as sns # for plotting
from sklearn import datasets
from sklearn import preprocessing
from sklearn import linear_model
from sklearn import model_selection as ms


# In[6]:


cancer_data = datasets.load_breast_cancer()
y = cancer_data.target # Training labels ('malignant = 0', 'benign = 1')
X = cancer_data.data # 30 attributes; https://scikit-learn.org/stable/datasets/index.html#breast-cancer-dataset
X = preprocessing.scale(X) # scale each data attribute to zero-mean and unit variance 


# In[24]:


# Plot the first 2 attributes of training points
sns.scatterplot(X[:, 0], X[:, 1], hue=y)
plt.xlabel('Tumor Radius')
plt.ylabel('Tumor Texture')
plt.grid(True)
plt.show()


# In[30]:


alpha = np.arange(1e-15,1,0.005) # Range of hyperparameter values 1E-15 to 1 by 0.005
val_scores = np.zeros((len(alpha),1)) # Initialize validation score for each alpha value

for i in range(len(alpha)): # for each alpha value
    # Set up SVM with hinge loss and l2 norm regularization
    model = linear_model.SGDClassifier(loss='hinge', penalty='l2', alpha=alpha[i])
    # Calculate cross validation scores for 5-fold cross-validation
    score = ms.cross_val_score(model, X, y, cv=5)
    val_scores[i] = score.mean() # Calculate mean of the 5 scores


# In[31]:


# Plot how cross-validation score changes with alpha
plt.plot(alpha,val_scores)
plt.xlim(0,1)
plt.xlabel('alpha')
plt.ylabel('Mean Cross-Validation Accuracy')
plt.grid(True)
plt.show()


# In[47]:


# Determine the alpha that maximizes the cross-validation score
ind = np.argmax(val_scores)
alpha_star = alpha[ind]
print('alpha_star =', alpha_star)

plt.plot(alpha,val_scores)
plt.plot(np.ones(11)*alpha_star,np.arange(0,1.1,0.1),'--r')
plt.xlim(0,1)
plt.ylim(0.94,0.98)
plt.xlabel('alpha')
plt.ylabel('Mean Cross-Validation Accuracy')
plt.grid(True)
plt.show()


# In[67]:


# Train model with alpha_star
model_star = linear_model.SGDClassifier(loss='hinge', penalty='l2', alpha=alpha_star)
model_trained = model_star.fit(X,y)
print('Training Accuracy =', model_trained.score(X,y))
# Training Accuracy = 0.9806678383128296


# In[62]:


# Plot decision boundary of trained model
slope = model_trained.coef_[0,1]/-model_trained.coef_[0,0]
x1 = np.arange(-10,10,0.5)
y1 = slope*x1
sns.scatterplot(X[:, 0], X[:, 1], hue=y)
plt.plot(x1,y1,'--k')
plt.xlim(-4,4)
plt.ylim(-6,6)
plt.xlabel('Tumor Radius')
plt.ylabel('Tumor Texture')
plt.grid(True)
plt.show()